test/index.js

```diff
@@ -105,10 +109,129 @@ describe('KeyringController', function () {
       await keyringController.persistAllKeyrings();
 
       const { vault } = keyringController.store.getState();
-      const keyrings = await mockEncryptor.decrypt(password, vault);
+      const keyrings = await mockEncryptor.decrypt(PASSWORD, vault);
       expect(keyrings.indexOf(unsupportedKeyring) > -1).toBe(true);
       expect(keyrings).toHaveLength(2);
     });
+
+    describe('when `cacheEncryptionKey` is enabled', function () {
+      it('should save an up to date encryption salt to the `memStore` when `password` is unset and `encryptionKey` is set', async function () {
+        keyringController.password = undefined;
+        keyringController.cacheEncryptionKey = true;
+        const vaultEncryptionKey = '🔑';
+        const vaultEncryptionSalt = '🧂';
+        const vault = JSON.stringify({ salt: vaultEncryptionSalt });
+        keyringController.store.updateState({ vault });
+
+        expect(keyringController.memStore.getState().encryptionKey).toBeNull();
+        expect(
+          keyringController.memStore.getState().encryptionSalt,
+        ).toBeUndefined();
+
+        await keyringController.unlockKeyrings(
+          undefined,
+          vaultEncryptionKey,
+          vaultEncryptionSalt,
+        );
+
+        expect(keyringController.memStore.getState().encryptionKey).toBe(
+          vaultEncryptionKey,
+        );
+        expect(keyringController.memStore.getState().encryptionSalt).toBe(
+          vaultEncryptionSalt,
+        );
+
+        const response = await keyringController.persistAllKeyrings();
+
+        expect(response).toBe(true);
+        expect(keyringController.memStore.getState().encryptionKey).toBe(
+          vaultEncryptionKey,
+        );
+        expect(keyringController.memStore.getState().encryptionSalt).toBe(
+          vaultEncryptionSalt,
+        );
+      });
+
+      it('should save an up to date encryption salt to the `memStore` when `password` is set through `createNewVaultAndKeychain`', async function () {
+        keyringController.cacheEncryptionKey = true;
+        const vaultEncryptionKey = '🔑';
+        const vaultEncryptionSalt = MOCK_HEX;
+        const vault = JSON.stringify({ salt: vaultEncryptionSalt });
+        keyringController.store.updateState({ vault });
+
+        await keyringController.createNewVaultAndKeychain(PASSWORD);
+
+        expect(keyringController.memStore.getState().encryptionKey).toBe(
+          MOCK_HARDCODED_KEY,
+        );
+        expect(keyringController.memStore.getState().encryptionSalt).toBe(
+          MOCK_HEX,
+        );
+
+        await keyringController.unlockKeyrings(
+          undefined,
+          vaultEncryptionKey,
+          vaultEncryptionSalt,
+        );
+
+        expect(keyringController.memStore.getState().encryptionKey).toBe(
+          vaultEncryptionKey,
+        );
+        expect(keyringController.memStore.getState().encryptionSalt).toBe(
+          vaultEncryptionSalt,
+        );
```

Comment:
Nit: The extra `unlockKeyrings` step in these tests makes it hard to tell what they're trying to assert. These two tests would be simpler without that step, and without the associated vault state.

It's a little confusing as well because with a real encryptor, the `unlockKeyrings` step would fail to decrypt the vault because it's using the wrong key. This test shows a key and a password that don't match being used to decrypt the same vault, which should be impossible with a working decryptor.

```suggestion
        await keyringController.createNewVaultAndKeychain(PASSWORD);
        expect(keyringController.memStore.getState().encryptionKey).toBe(
          MOCK_HARDCODED_KEY,
        );
        expect(keyringController.memStore.getState().encryptionSalt).toBe(
          MOCK_HEX,
        );
```

The same simplification can be made to the test immediately following this one as well